### ISSUE TITLE
Think about when to return an Arr or an array

### ISSUE THREAD

**Issue by cliffckerr** [starsimhub/starsim#508]
Should `sim.people.age + 1` return a `FloatArr` or a Numpy array? Probably `FloatArr`, with `sim.people.age[:] + 1` returning an array?

**Comment by robynstuart** [comment]
Related, I encountered this today:
```
(Pdb) >? unborn_uids
uids([10057, 10156])
(Pdb) >? unborn_uids += [203]
(Pdb) >? unborn_uids
uids([10260, 10359])
```

What's the recommended syntax for adding a UID to a list? `unborn_uids += ss.uids([203])` also just added 203 to the individual UIDS, rather than adding a new UID for agent 203 like I was hoping.

**Comment by cliffckerr** [comment]
Good question -- I suggest making it behave more like a list:
- `uids1 + uids2 = cat(uids1, uids2)`
- all other arithmetic operations are disallowed
- `uids.to_numpy()` is renamed `uids.values`

I don't know if it will be possible to get in-place operations to work, e.g. `uids1 += uids2` ... I get:
```py
import starsim as ss
u1 = ss.uids([1,2,3])
u1.resize((4,1), refcheck=False)

Traceback (most recent call last):
  Cell In[7], line 1
    u1.resize((4,1), refcheck=False)

ValueError: cannot resize this array: it does not own its data
```
@RomeshA @pausz if you have any ideas here, they'd be much appreciated! My two thoughts are:
- `uids` is constructed as `np.array(data).view(uids)`, so if we could temporarily "unview" it, resize, and then "review" it, that could work? Maybe?
- Whereas `Arr` inherits from `np.lib.mixins.NDArrayOperatorsMixin`, `uids` inherits directly from `np.ndarray`. It is _so close_ to working, but this might be a fundamental limitation. Especially if we're making `uids` less array-like and more list-like, this direct inheritance might be unnecessary.
- It's not the end of the world if in-place operations fail, and one could argue that it's actually helpful to clarify when a memory copy is taking place. Since everything else works, maybe it's fine to just leave this.

**Comment by kaa2102** [comment]
@robynstuart @cliffckerr I encountered similar issues with trying to add/concat/manipulate uid arrays. In aligning Births and Deaths, I used sim.people.grow() to create new baby uids and connect them to mothers with new states ss.Arr() pf type List dtype=[ ]. I had trouble trying to manipulate or concatenate the uid arrays so I worked with lists of uids instead (np.array(arr1), arr1.tolist(), np.hstack([arr1,arr2]).

**Comment by RomeshA** [comment]
> Should `sim.people.age + 1` return a `FloatArr` or a Numpy array? Probably `FloatArr`, with `sim.people.age[:] + 1` returning an array?

For this one, it would be _nice_ if it returned a `FloatArr`, however, when I tested something similar this slowed the code down a lot, because `Arr.asnew()` is relatively heavy (since you need to index the new `raw` array to insert the new items - `new.raw[new.auids] = arr`) and inside calculations, there could be many such operations, which results in the creation of a large number of intermediate `Arr` instances that we don't actually care about. That was also part of the reason that the original `States` implementation was slower than the current implementation, because it wrapped the intermediate results in our custom container to allow indexing by UIDs. So I'd suggest that we just leave it as it is, unless we find a way to make `asnew()` faster

**Comment by RomeshA** [comment]
> Related, I encountered this today:
> 
> ```
> (Pdb) >? unborn_uids
> uids([10057, 10156])
> (Pdb) >? unborn_uids += [203]
> (Pdb) >? unborn_uids
> uids([10260, 10359])
> ```
> 
> What's the recommended syntax for adding a UID to a list? `unborn_uids += ss.uids([203])` also just added 203 to the individual UIDS, rather than adding a new UID for agent 203 like I was hoping.

`ss.uids` operators now perform set operations, so the recommended way would be to think of this as taking the union of the two sets of UIDs, with the `|` operator. So it would be `unborn_uids = unborn_uids | [203]` (actually `unborn_uids | 203` works as well, if you were only adding one)

**Comment by cliffckerr** [comment]
Closed by #670
